guestcluster.go

```diff
@@ -0,0 +1,163 @@
+package guestcluster
+
+import (
+	"context"
+
+	"github.com/giantswarm/apiextensions/pkg/clientset/versioned"
+	"github.com/giantswarm/certs"
+	"github.com/giantswarm/helmclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/operatorkit/client/k8srestconfig"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+// Config represents the configuration used to create a new guest cluster service.
+type Config struct {
+	CertsSearcher certs.Interface
+	Logger        micrologger.Logger
+
+	CertID certs.Cert
+}
+
+// GuestCluster provides functionality for connecting to guest clusters.
+type GuestCluster struct {
+	certsSearcher certs.Interface
+	logger        micrologger.Logger
+
+	certID certs.Cert
+}
+
+// New creates a new guest cluster service.
+func New(config Config) (*GuestCluster, error) {
+	if config.CertsSearcher == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.CertsSearcher must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	if config.CertID == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.CertID must not be empty", config)
+	}
+
+	g := &GuestCluster{
+		certsSearcher: config.CertsSearcher,
+		logger:        config.Logger,
+
+		certID: config.CertID,
+	}
+
+	return g, nil
+}
+
+// NewG8sClient returns a generated clientset for the specified guest cluster.
+func (g *GuestCluster) NewG8sClient(ctx context.Context, clusterID, apiDomain string) (versioned.Interface, error) {
+	g.logger.LogCtx(ctx, "level", "debug", "message", "creating G8s client for the guest cluster")
+
+	restConfig, err := g.newRestConfig(ctx, clusterID, apiDomain)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	g8sClient, err := versioned.NewForConfig(restConfig)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	g.logger.LogCtx(ctx, "level", "debug", "message", "created G8s client for the guest cluster")
+
+	return g8sClient, nil
+}
+
+// NewHelmClient returns a Helm client for the specified guest cluster.
+func (g *GuestCluster) NewHelmClient(ctx context.Context, clusterID, apiDomain string) (helmclient.Interface, error) {
+	g.logger.LogCtx(ctx, "level", "debug", "message", "creating Helm client for the guest cluster")
+
+	restConfig, err := g.newRestConfig(ctx, clusterID, apiDomain)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	k8sClient, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	c := helmclient.Config{
+		K8sClient: k8sClient,
+		Logger:    g.logger,
+
+		RestConfig:      restConfig,
+		TillerNamespace: tillerDefaultNamespace,
```

Comment:
For managed services we want to use the Tiller we create in `giantswarm`. In other contexts we might want to use Tiller in `kube-system`. Should we make this configurable?

Reply:
Is this actually a separation of host and guest cluster access? `guestcluster` should only be about guest clusters and the host cluster magic should maybe be somewhere else? Uncertain. I still do not have a clear picture on the namespace difference.

Reply:
For context we deploy a separate Tiller pod in `giantswarm` for isolation and to avoid conflicting with customer workloads. Since it's common practice to install Tiller in `kube-system`.

I agree `guestcluster` should only care about guest clusters. So after all, I think this is fine.

Reply:
Is it that we run two tillers in the guest clusters in two different namespaces? I would just make the tiller namespace configurable. Either during service initialization or as function parameter. The latter is a bit odd because I like the consistent interface here. :D

Reply:
We run tilller in the giantswarm namespace. There may also be a customer installed tiller in kube-system. Making the tiller namespace configurable during the service installation makes sense to me.

Reply:
I see. Cool. Will do.

Reply:
I made this configurable.